core/include/vecmem/containers/impl/device_vector.ipp

```diff
@@ -225,6 +227,50 @@ VECMEM_HOST_AND_DEVICE auto device_vector<TYPE>::push_back(
     return index;
 }
 
+template <typename TYPE>
+VECMEM_HOST_AND_DEVICE auto device_vector<TYPE>::reserve(size_type n)
+    -> size_type {
+    // This can only be done on a resizable vector.
+    assert(m_size != nullptr);
+
+#if defined(__cpp_lib_is_implicit_lifetime) && \
+    __cpp_lib_is_implicit_lifetime >= 202302L
+    static_assert(std::is_implicit_lifetyime_v<TYPE>,
+                  "Type `T` in `device_vector<T>::reserve` is not an implicit "
+                  "lifetype type (according to the C++23-compliant "
+                  "definition), so slots cannot be safely reserved.");
+#else
+    static_assert(details::is_implicit_lifetime_v<TYPE>,
+                  "Type `T` in `device_vector<T>::reserve` is not an implicit "
+                  "lifetype type (according to a non-C++23-compliant "
+                  "definition), so slots cannot be safely reserved.");
```

Comment:
We don't actually require this for device types at the moment. :thinking: This is why the `device_vector<T>::construct(...)` function exists. I was expecting that you'd make use of it here.

Making it possible to use types that have some simple constructors (for setting default values on variables for instance), would be highly desirable in my mind.

Reply:
Right, but there's a problem with that strategy: if you want to reserve, say, 1024 elements for a whole block of threads, the constructor approach requires the calling thread to loop through a constructor 1024 times which could be potentially super slow.

The advantage of using implicit lifetime types like this is that the object is in a well-defined (but unknown) stage the second the memory is allocated, so serial construction is not necessary.

The unsafe method allows the client to construct the objects in parallel for performance.

In principle I think having a set of functions which enlarges a vector by $n$ elements and the constructs them in place also has a use, but it's a different use from this.

Reply:
:+1: